hidimstat/desparsified_lasso.py

```diff
@@ -1,79 +1,133 @@
 import numpy as np
+from numpy.linalg import multi_dot
 import scipy.stats as st
+from scipy.linalg import inv
 from joblib import Parallel, delayed
-from sklearn.linear_model import Lasso, LassoCV
+from sklearn.utils.validation import check_memory
+from sklearn.preprocessing import StandardScaler
+from sklearn.linear_model import Lasso
 
-from .noise_std import reid
+from .noise_std import reid, group_reid
+from .stat_tools import sf_and_cdf_from_pval_and_sign
 
 
-def _compute_residuals(X, column_index, alpha=None, Gram=None, max_iter=5000,
+def _compute_all_residuals(X, alphas, gram=None, max_iter=5000, tol=1e-3,
+                           method='lasso', c=0.01, n_jobs=1, verbose=0):
+    """Nodewise Lasso. Compute all the residuals: regressing each column of the
+    design matrix against the other columns"""
+
+    n_samples, n_features = X.shape
+
+    results = \
+        Parallel(n_jobs=n_jobs, verbose=verbose)(
+            delayed(_compute_residuals)
+                (X=X,
+                 column_index=i,
+                 alpha=alphas[i],
+                 gram=gram,
+                 max_iter=max_iter,
+                 tol=tol,
+                 method=method,
+                 c=c)
+            for i in range(n_features))
+
+    results = np.asarray(results)
+    Z = np.stack(results[:, 0], axis=1)
+    omega_diag = np.stack(results[:, 1])
+
+    return Z, omega_diag
+
+
+def _compute_residuals(X, column_index, alpha=None, gram=None, max_iter=5000,
                        tol=1e-3, method='lasso', c=0.01):
-    """Nodewise Lasso"""
+    """Compute the residuals of the regression of a given column of the
+    design matrix against the other columns"""
 
     n_samples, n_features = X.shape
     i = column_index
 
     X_new = np.delete(X, i, axis=1)
     y = np.copy(X[:, i])
 
-    if method == 'lasso' and Gram is None:
-        Gram_loc = np.dot(X_new.T, X_new)
-    elif method == 'lasso' and Gram is not None:
-        Gram_loc = np.delete(np.delete(Gram, i, axis=0), i, axis=1)
+    if method != 'lasso':
+        ValueError("The only regression method available is 'lasso'")
+
+    if gram is None:
+        gram_loc = False
+    else:
+        gram_loc = np.delete(np.delete(gram, i, axis=0), i, axis=1)
 
-    if method == 'lasso' and alpha is None:
+    if alpha is None:
         k = c * (1. / n_samples)
```

Comment:
if you can avoid single letter variable names it's much appreciated. eg alpha_max_fraction for example.